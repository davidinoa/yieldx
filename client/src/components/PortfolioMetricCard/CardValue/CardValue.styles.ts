import styled from "styled-components";
import { mediaQuery } from "../../../theme/breakpoints.style";

export const Root = styled.div`
  line-height: normal;
  font-size: 1.5rem;

  ${mediaQuery("xxl")} {
    font-size: 1.95rem;
  }

  sup {
    font-size: 1rem;
    padding: 0.25rem;

    ${mediaQuery("xxl")} {
      font-size: 1.15rem;
    }
  }

  span {
    color: var(--subtext-3);
    font-size: 12px;
    line-height: 15px;
  }

  .allocation-separator {
    height: 23px;
    width: 2px;
    background: var(--border-new);
  }
`;
