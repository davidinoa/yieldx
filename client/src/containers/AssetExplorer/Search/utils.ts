interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
