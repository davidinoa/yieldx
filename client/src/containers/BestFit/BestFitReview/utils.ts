import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
