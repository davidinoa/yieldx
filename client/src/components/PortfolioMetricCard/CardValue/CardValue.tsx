import React from "react";
import Decimal from "../../../util/Decimal";
import { Root } from "./CardValue.styles";

interface Value {
  number: number;
  format: "%" | "$" | "bps";
  helperText?: string;
}

interface Props {
  value: Value;
  secondValue?: Value;
}

function renderValue({ number, format, helperText }: Value) {
  const formattedValue = Decimal.format(
    format === "bps" ? Math.round(number) : number
  );
  return (
    <>
      {format === "$" && `$${formattedValue.split(".")[0]}`}
      {format === "%" && formattedValue}
      {format === "bps" && `${formattedValue.split(".")[0]}`}
      {format === "%" && <sup>%</sup>}
      {format === "bps" && <sup>bps</sup>}
      {format === "$" && <sup>.{formattedValue.split(".")[1]}</sup>}
      {helperText && <span>{`/${helperText}`}</span>}
    </>
  );
}

export default function CardValue({ value, secondValue }: Props) {
  return (
    <Root>
      {secondValue === undefined && renderValue(value)}
      {secondValue !== undefined && (
        <div className="d-flex justify-content-between align-items-center">
          <div>{renderValue(value)}</div>
          <div className="allocation-separator" />
          <div>{renderValue(secondValue)}</div>
        </div>
      )}
    </Root>
  );
}
