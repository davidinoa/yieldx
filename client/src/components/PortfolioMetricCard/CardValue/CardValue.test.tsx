import React from "react";
import { render, screen } from "../../../test/utils";
import CardValue from "./CardValue";

describe("CardValue", () => {
  test("Dollar format snapshot", () => {
    const { asFragment } = render(
      <CardValue
        value={{
          format: "$",
          number: 5_000.25
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("Percent format snapshot", () => {
    const { asFragment } = render(
      <CardValue
        value={{
          format: "%",
          number: 12.25
        }}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });
  test("BPS format snapshot", () => {
    // bps should be rounded
    const { asFragment } = render(
      <CardValue
        value={{
          format: "bps",
          number: 250.25
        }}
      />
    );
    expect(screen.queryByText(/250/)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  test("Helper text snapshot", () => {
    const { asFragment } = render(
      <CardValue
        value={{
          format: "$",
          number: 5_000.25,
          helperText: "monthly"
        }}
      />
    );
    expect(screen.queryByText(/\/monthly/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  test("Should display 2 values if secondValue not undefined", () => {
    const { asFragment } = render(
      <CardValue
        value={{
          format: "%",
          number: 3,
          helperText: "min"
        }}
        secondValue={{
          format: "%",
          number: 5.2,
          helperText: "max"
        }}
      />
    );
    expect(screen.queryByText(/\/min/i)).toBeInTheDocument();
    expect(screen.queryByText(/\/max/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
