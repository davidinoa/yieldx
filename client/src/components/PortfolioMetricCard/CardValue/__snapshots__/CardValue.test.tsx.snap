// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardValue BPS format snapshot 1`] = `
<DocumentFragment>
  .c0 {
  line-height: normal;
  font-size: 1.5rem;
}

.c0 sup {
  font-size: 1rem;
  padding: 0.25rem;
}

.c0 span {
  color: var(--subtext-3);
  font-size: 12px;
  line-height: 15px;
}

.c0 .allocation-separator {
  height: 23px;
  width: 2px;
  background: var(--border-new);
}

@media (min-width:1600px) {
  .c0 {
    font-size: 1.95rem;
  }
}

@media (min-width:1600px) {
  .c0 sup {
    font-size: 1.15rem;
  }
}

<div
    class="c0"
  >
    250
    <sup>
      bps
    </sup>
  </div>
</DocumentFragment>
`;

exports[`CardValue Dollar format snapshot 1`] = `
<DocumentFragment>
  .c0 {
  line-height: normal;
  font-size: 1.5rem;
}

.c0 sup {
  font-size: 1rem;
  padding: 0.25rem;
}

.c0 span {
  color: var(--subtext-3);
  font-size: 12px;
  line-height: 15px;
}

.c0 .allocation-separator {
  height: 23px;
  width: 2px;
  background: var(--border-new);
}

@media (min-width:1600px) {
  .c0 {
    font-size: 1.95rem;
  }
}

@media (min-width:1600px) {
  .c0 sup {
    font-size: 1.15rem;
  }
}

<div
    class="c0"
  >
    $5,000
    <sup>
      .25
    </sup>
  </div>
</DocumentFragment>
`;

exports[`CardValue Helper text snapshot 1`] = `
<DocumentFragment>
  .c0 {
  line-height: normal;
  font-size: 1.5rem;
}

.c0 sup {
  font-size: 1rem;
  padding: 0.25rem;
}

.c0 span {
  color: var(--subtext-3);
  font-size: 12px;
  line-height: 15px;
}

.c0 .allocation-separator {
  height: 23px;
  width: 2px;
  background: var(--border-new);
}

@media (min-width:1600px) {
  .c0 {
    font-size: 1.95rem;
  }
}

@media (min-width:1600px) {
  .c0 sup {
    font-size: 1.15rem;
  }
}

<div
    class="c0"
  >
    $5,000
    <sup>
      .25
    </sup>
    <span>
      /monthly
    </span>
  </div>
</DocumentFragment>
`;

exports[`CardValue Percent format snapshot 1`] = `
<DocumentFragment>
  .c0 {
  line-height: normal;
  font-size: 1.5rem;
}

.c0 sup {
  font-size: 1rem;
  padding: 0.25rem;
}

.c0 span {
  color: var(--subtext-3);
  font-size: 12px;
  line-height: 15px;
}

.c0 .allocation-separator {
  height: 23px;
  width: 2px;
  background: var(--border-new);
}

@media (min-width:1600px) {
  .c0 {
    font-size: 1.95rem;
  }
}

@media (min-width:1600px) {
  .c0 sup {
    font-size: 1.15rem;
  }
}

<div
    class="c0"
  >
    12.25
    <sup>
      %
    </sup>
  </div>
</DocumentFragment>
`;

exports[`CardValue Should display 2 values if secondValue not undefined 1`] = `
<DocumentFragment>
  .c0 {
  line-height: normal;
  font-size: 1.5rem;
}

.c0 sup {
  font-size: 1rem;
  padding: 0.25rem;
}

.c0 span {
  color: var(--subtext-3);
  font-size: 12px;
  line-height: 15px;
}

.c0 .allocation-separator {
  height: 23px;
  width: 2px;
  background: var(--border-new);
}

@media (min-width:1600px) {
  .c0 {
    font-size: 1.95rem;
  }
}

@media (min-width:1600px) {
  .c0 sup {
    font-size: 1.15rem;
  }
}

<div
    class="c0"
  >
    <div
      class="d-flex justify-content-between align-items-center"
    >
      <div>
        3.00
        <sup>
          %
        </sup>
        <span>
          /min
        </span>
      </div>
      <div
        class="allocation-separator"
      />
      <div>
        5.20
        <sup>
          %
        </sup>
        <span>
          /max
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
